src/ucs/type/spinlock.h

```diff
@@ -62,35 +89,74 @@ static inline ucs_status_t ucs_spinlock_destroy(ucs_spinlock_t *lock)
     return UCS_OK;
 }
 
-static inline int ucs_spin_is_owner(ucs_spinlock_t *lock, pthread_t self)
+static inline ucs_status_t ucs_recursive_spinlock_destroy(ucs_recursive_spinlock_t *lock)
+{
+    if (lock->count != 0) {
+        return UCS_ERR_BUSY;
+    }
+
+    return ucs_spinlock_destroy(&lock->super);
+}
+
+static inline int ucs_recursive_spin_is_owner(ucs_recursive_spinlock_t *lock, pthread_t self)
 {
     return lock->owner == self;
 }
 
 static inline void ucs_spin_lock(ucs_spinlock_t *lock)
+{
+#if UCS_ENABLE_ASSERT
+    ucs_assert(lock->owner == UCS_SPINLOCK_OWNER_NULL);
+#endif
+
+    pthread_spin_lock(&lock->lock);
+
+#if UCS_ENABLE_ASSERT
+    lock->owner = pthread_self();
+#endif
+}
+
+static inline void ucs_recursive_spin_lock(ucs_recursive_spinlock_t *lock)
 {
     pthread_t self = pthread_self();
 
-    if (ucs_spin_is_owner(lock, self)) {
+    if (ucs_recursive_spin_is_owner(lock, self)) {
         ++lock->count;
         return;
     }
 
-    pthread_spin_lock(&lock->lock);
+    ucs_spin_lock(&lock->super);
     lock->owner = self;
     ++lock->count;
 }
 
-static inline int ucs_spin_trylock(ucs_spinlock_t *lock)
+static inline int ucs_spin_try_lock(ucs_spinlock_t *lock)
+{
+#if UCS_ENABLE_ASSERT
```

Comment:
no need to wrap ucs_assert inside ENABLE_ASSERT

Reply:
@yosefe Turns out wrapping is needed... otherwise `ucs_spinlock_t` doesn't even contain "owner" member, and I get:

````
/__w/1/s/build/../src/ucs/type/spinlock.h:110:20: error: 'ucs_spinlock_t' has no member named 'owner'
     ucs_assert(lock->owner == UCS_SPINLOCK_OWNER_NULL);
````

Reply:
there is probably some other problem since ucs_assert is already not doing anything for release build

Reply:
Not sure it's THE problem, but it's A problem: `debug/assert.h` is under `noinst_HEADERS`, so it's not installed - and `ucs_assert()` is not available to the installed `spinlock.h`. 

@yosefe - would you agree to move `debug/assert.h` into `nobase_dist_libucs_la_HEADERS` ?

Reply:
i'm afraid it's not possible to install headers which use assert, because of config.h dependency, which cannot be installed